src/test/spec/unified_runner/matcher.rs

```diff
@@ -86,30 +86,20 @@ fn type_matches(types: &Bson, actual: &Bson) -> bool {
             "maxKey" => actual.element_type() == ElementType::MaxKey,
             other => panic!("unrecognized type: {}", other),
         },
-        Bson::Int32(n) => match n {
-            1 => actual.element_type() == ElementType::Double,
-            2 => actual.element_type() == ElementType::String,
-            3 => actual.element_type() == ElementType::EmbeddedDocument,
-            4 => actual.element_type() == ElementType::Array,
-            5 => actual.element_type() == ElementType::Binary,
-            6 => actual.element_type() == ElementType::Undefined,
-            7 => actual.element_type() == ElementType::ObjectId,
-            8 => actual.element_type() == ElementType::Boolean,
-            9 => actual.element_type() == ElementType::DateTime,
-            10 => actual.element_type() == ElementType::Null,
-            11 => actual.element_type() == ElementType::RegularExpression,
-            12 => actual.element_type() == ElementType::DbPointer,
-            13 => actual.element_type() == ElementType::JavaScriptCode,
-            14 => actual.element_type() == ElementType::Symbol,
-            15 => actual.element_type() == ElementType::JavaScriptCodeWithScope,
-            16 => actual.element_type() == ElementType::Int32,
-            17 => actual.element_type() == ElementType::Timestamp,
-            18 => actual.element_type() == ElementType::Int64,
-            19 => actual.element_type() == ElementType::Decimal128,
-            -1 => actual.element_type() == ElementType::MinKey,
-            127 => actual.element_type() == ElementType::MaxKey,
-            other => panic!("unrecognized type: {}", other),
-        },
+        Bson::Int32(n) => {
+            let n = *n;
+            // -1 cannot be converted to a u8
+            if n == -1 {
```

Comment:
Now that I'm looking at this again, I'm surprised by this check. Is there any reason why the unified spec test format can't just use 0xFF (or 255, if hex isn't supported)? I don't think it should make anything less difficult for languages without strict integer types, and it will make things easier for languages like us that do.

Reply:
For context, -1 is only used for consistency with https://docs.mongodb.com/manual/reference/operator/query/type/. I imagine u8 is being used here because the rest of the values happen to fall in the range of 1...127, but it'd probably be more straightforward to use a signed 8-bit integer if that's possible.

Reply:
Unfortunately we can't change the BSON library itself, as that could break user code. I'm honestly pretty surprised that a signed integer is used anywhere; I've never thought of BSON type tags as signed integers, and I'm not really sure why anyone would ever prefer to do that.

Reply:
FWIW in Swift we also used a `UInt8` as the raw type for for our `BSONType` enum, with 0xFF representing the MinKey case: https://github.com/mongodb/swift-bson/blob/b8d953ae93951e7bee6cde4f7cc6774211378982/Sources/SwiftBSON/BSONValue.swift#L95

we expose this via a `bsonType` property on each BSON value.

this is for consistency with the BSON spec itself and the values it uses for encoded type prefixes. I know libbson similarly has a [`bson_type_t`](http://mongoc.org/libbson/current/bson_type_t.html) enum which uses 0xFF for MinKey.

we could work around this by special-casing as Isabel does here, but I'm curious what the benefit is of going with consistency with the query operator over with how a number of drivers already represent BSON type via integer?

Reply:
I didn't that this was interfacing with your BSON library, and it never occurred to me that the numbers in https://docs.mongodb.com/manual/reference/operator/query/type/ were related to the type codes in http://bsonspec.org/spec.html. In that case, I'd rather remove support for a numeric code in `$$type` and stick to strings.

Reply:
That sounds good to me! It will make the spec tests more human-readable for sure

Reply:
https://github.com/mongodb/specifications/pull/846/commits/5480885e819e8affe154b38954e3ae09f36caa5e

Reply:
Removed the `Int32` case here.